### PR TITLE
Add new lint `upper_case_acronyms`

### LINKED ISSUE #1335: Naming convention
At work I have a VS plugin warning about names such as `TOMLParser`. [According to Rust's naming conventions, this should be `TomlParser`.](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md#fine-points)
We should warn on more that 2 consecutive capital letters in a name.

### PR DESCRIPTION
**Pull request by matsujika**
Close #1335
I need some reviews on the English sentences because I feel they're messed up. ;)

changelog: Add new lint `upper_case_acronyms`

### CODE REVIEW

**Comment by mikerite** [comment]
`CCHelper` → `CcHelper` and `SAOnly` → `SaOnly` looks pretty weird to me. In .Net, at least, two letter acronyms are in uppercase. See [.Net documenation - Capitalization Conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions).  The issue talks about VS plugins so those are probably the rules they are using. I tried to find some Rust documentation on this but I didn't find anything.

Also, I think this might have to be pedantic, as it will generate excessive warnings in some projects.

**Comment by matsujika** [comment]
> CCHelper → CcHelper and SAOnly → SaOnly looks pretty weird to me. In .Net, at least, two letter acronyms are in uppercase. See .Net documenation - Capitalization Conventions. The issue talks about VS plugins so those are probably the rules they are using. I tried to find some Rust documentation on capitalization conventions of acronyms but I didn't find anything.

I have the same spontaneous feeling on two-letter acronyms but IMO they aren't exceptions in Rust given [IoSlice](https://doc.rust-lang.org/std/io/struct.IoSlice.html), [TypeId](https://doc.rust-lang.org/std/any/struct.TypeId.html) and [GlObject](https://docs.rs/glium/0.29.0/glium/trait.GlObject.html).

> Also, I think this might have to be pedantic, as it will generate excessive warnings in some projects.

If two-letter acronyms are exceptions pedantic is fair as the lint will be false-positive (including ID and IO) prone. If not I'd prefer this lint to be deny-by-default since the lint should be as broad as possible, considering the naming convention on acronyms are not intuitive for those who came from other languages.

**Comment by mikerite** [comment]
The `IoSlice` example convinced me that two letter acronyms aren't an exception. 

I'll let the reviewer decide on whether this should be pedantic or something else. I tested it on a bunch of crates and there weren't that many warnings.

The one issue I found is that it lints on globals created with lazy_static.

```rust
use lazy_static::lazy_static;

lazy_static! {
    /// This is an example for using doc comment attributes
    static ref EXAMPLE: u8 = 42;
}

fn main() {}
```

```
error: name `EXAMPLE` may contain capitalized acronym
  --> src/main.rs:11:16
   |
11 |     static ref EXAMPLE: u8 = 42;
   |                ^^^^^^^ help: consider making the acronym lowercase expect the initial letter: `Example`
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
```
There probably needs to be a macro check somewhere.

I was also wondering if we should call the lint `upper-case-acronyms` to be consistent with the [non-upper-case-globals](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#non-upper-case-globals) rust lint.

**Comment by matsujika** [comment]
> I'll let the reviewer decide on whether this should be pedantic or something else. I tested it on a bunch of crates and there weren't that many warnings.

Got it. I guess I need opinions on its lint category.

> The one issue I found is that it lints on globals created with lazy_static.

Very nice catch! `lazy_static` expands `EXAMPLE` to a struct, leading to being linted. Perhaps this lint should ignore macros.
I feel `upper-case-acronyms` is nicer than `capitalized-acronyms`, I'll rename it to that.

I'll make the changes next week as I have a lot to do right now.

File: clippy_lints/src/capitalized_acronyms.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            &format!("name `{}` may contain capitalized acronym", ident),
+            "consider making the acronym lowercase expect the initial letter",
+            corrected,
+            Applicability::MaybeIncorrect,
+        )
+    }
+}
+
+impl EarlyLintPass for CapitalizedAcronyms {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, it: &Item) {
```
Comment (flip1995) [inline clippy_lints/src/capitalized_acronyms.rs:73]: We should not lint `in_external_macro`s.

File: clippy_lints/src/capitalized_acronyms.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+use rustc_span::symbol::Ident;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for camel case name containing a capitalized acronym.
+    ///
+    /// **Why is this bad?** In CamelCase, acronyms count as one word.
+    /// See [naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#general-conventions-[rfc-#430])
+    /// for more.
+    ///
+    /// **Known problems:** None.
```
Comment (flip1995) [inline clippy_lints/src/capitalized_acronyms.rs:16]: I think one problem is, that if two acronyms are chained together they are both lowercased by the suggestion. E.g.:

```
GCCLLVMSomething -> GccllvmSomething
```

Can you also add a test for this, please?

File: clippy_lints/src/capitalized_acronyms.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    let span = ident.span;
+    let ident = &ident.as_str();
+    let corrected = correct_ident(ident);
+    if ident != &corrected {
+        span_lint_and_sugg(
+            cx,
+            CAPITALIZED_ACRONYMS,
+            span,
+            &format!("name `{}` may contain capitalized acronym", ident),
+            "consider making the acronym lowercase expect the initial letter",
```
Comment (flip1995) [inline clippy_lints/src/capitalized_acronyms.rs:65]:
```suggestion
            "consider making the acronym lowercase, except the initial letter",
```

File: clippy_lints/src/capitalized_acronyms.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+fn check_ident(cx: &EarlyContext<'_>, ident: &Ident) {
+    let span = ident.span;
+    let ident = &ident.as_str();
+    let corrected = correct_ident(ident);
+    if ident != &corrected {
+        span_lint_and_sugg(
+            cx,
+            CAPITALIZED_ACRONYMS,
+            span,
+            &format!("name `{}` may contain capitalized acronym", ident),
```
Comment (flip1995) [inline clippy_lints/src/capitalized_acronyms.rs:64]:
```suggestion
            &format!("name `{}` contains a capitalized acronym", ident),
```

File: tests/ui/complex_types.rs [hunk 5]
```diff
@@ -1,5 +1,10 @@
 #![warn(clippy::all)]
-#![allow(unused, clippy::needless_pass_by_value, clippy::vec_box)]
+#![allow(
+    unused,
+    clippy::needless_pass_by_value,
+    clippy::vec_box,
+    clippy::capitalized_acronyms
```
Comment (flip1995) [inline tests/ui/complex_types.rs:6]: I think in this case it would be better to change this in the tests.

**Review by flip1995 — COMMENTED** [review]
I think style is appropriate for this lint.


> I tried to find some Rust documentation on capitalization conventions of acronyms but I didn't find anything.


@mikerite  The API guidelines of rust mention this in the paragraph after the table, you can find here: https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case

@matsujika ^ Please link this in the link documentation instead, since it is more current.

File: clippy_lints/src/upper_case_acronyms.rs [hunk 6]
```diff
@@ -1,19 +1,21 @@
...
 declare_clippy_lint! {
     /// **What it does:** Checks for camel case name containing a capitalized acronym.
     ///
     /// **Why is this bad?** In CamelCase, acronyms count as one word.
-    /// See [naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#general-conventions-[rfc-#430])
+    /// See [naming conventions](https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case)
     /// for more.
     ///
-    /// **Known problems:** None.
+    /// **Known problems:** When two acronyms are contiguous, they are interpreted as a one acronym.
```
Comment (flip1995) [inline clippy_lints/src/upper_case_acronyms.rs:18]:
```suggestion
    /// **Known problems:** When two acronyms are contiguous, the lint can't tell where
    /// the first acronym ends and the second starts, so it suggests to lowercase all of
    /// the letters in the second acronym.
```

File: clippy_lints/src/upper_case_acronyms.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    /// **Example:**
+    ///
+    /// ```rust
+    /// struct HTTPResponse;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct HttpResponse;
+    /// ```
+    pub UPPER_CASE_ACRONYMS,
```
Comment (flip1995) [inline clippy_lints/src/upper_case_acronyms.rs:29]:
```suggestion
    pub UPPERCASE_ACRONYMS,
```

Uppercase is usually written as one word. Like in the [`char::to_uppercase`](https://doc.rust-lang.org/std/primitive.char.html#method.to_uppercase) method

**Review by flip1995 — COMMENTED** [review]
Two small things left

File: clippy_lints/src/upper_case_acronyms.rs (hunk 7 above)
Comment (matsujika) [reply clippy_lints/src/upper_case_acronyms.rs:31]: There's `non-upper-case-globals` lint in rustc and I suppose it's reasonable to make the name of this lint similar to that.
(It seems they wanted to go consistent with `non-camel-case-types` and `non_snake_case` when naming the rustc lint)

Comment (flip1995) [reply clippy_lints/src/upper_case_acronyms.rs:31]: Well, I find that inconsistent. But lets be consistently inconsistent and keep the name as it is 👍

**Review by flip1995 — APPROVED** [review]
Thanks! Please squash some of your commits, so we can merge this.

File: clippy_lints/src/lib.rs [hunk 8]
```diff
@@ -6,6 +6,7 @@
 #![feature(concat_idents)]
 #![feature(crate_visibility_modifier)]
 #![feature(drain_filter)]
+#![feature(split_inclusive)]
```
Comment (flip1995) [inline clippy_lints/src/lib.rs:9]:
```suggestion
```

This is stable now.

Comment (matsujika) [reply clippy_lints/src/lib.rs]: Happy to see it has been stabilized :)

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**